core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.view;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.util.PropertyUtil;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class ViewMetadata implements Serializable {
+  static final int SUPPORTED_VIEW_FORMAT_VERSION = 1;
+
+  public abstract int formatVersion();
+
+  public abstract String location();
+
+  public abstract Integer currentSchemaId();
+
+  public abstract List<Schema> schemas();
+
+  public abstract int currentVersionId();
+
+  public abstract List<ViewVersion> versions();
+
+  public abstract List<ViewHistoryEntry> history();
+
+  public abstract Map<String, String> properties();
+
+  public ViewVersion version(int versionId) {
+    return versionsById().get(versionId);
+  }
+
+  @Value.Lazy
```

Comment:
I think this should be `Derived` rather than `Lazy`. Lazy makes it more complicated and there's no real benefit to not just calculating this at construction.

Reply:
In that case, does it make sense to make this `Lazy` at all? They're not lazy or derived in `TableMetadata`:

```java
  public Schema schema() {
    return schemasById.get(currentSchemaId);
  }
```

Reply:
I'm ok removing `@Value.Lazy` here, since the only thing it buys us is that it doesn't re-compute the value from the underlying hash map, which is O(1) anyway